src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs

```diff
@@ -0,0 +1,396 @@
+
+        public static void AddScale(float a, float[] src, float[] dst, int count)
+        {
+            Contracts.AssertNonEmpty(src);
+            Contracts.Assert(0 < count && count <= src.Length);
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(count <= dst.Length);
+
+            AddScale(a, new Span<float>(src, 0, count), new Span<float>(dst, 0, count));
+        }
+
+        public static void AddScale(float a, float[] src, float[] dst, int dstOffset, int count)
+        {
+            Contracts.AssertNonEmpty(src);
+            Contracts.Assert(count <= src.Length);
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(0 <= dstOffset && dstOffset < dst.Length);
+            Contracts.Assert(0 < count && count <= dst.Length - dstOffset);
+
+            AddScale(a, new Span<float>(src, 0, count), new Span<float>(dst, dstOffset, count));
+        }
+
+        private static void AddScale(float a, Span<float> src, Span<float> dst)
+        {
+            if (Sse.IsSupported)
+            {
+                SseIntrinsics.AddScaleU(a, src, dst);
+            }
+            else
+            {
+                for (int i = 0; i < dst.Length; i++)
+                {
+                    dst[i] += a * src[i];
+                }
+            }
+        }
+
+        public static void AddScale(float a, float[] src, int[] indices, float[] dst, int count)
+        {
+            Contracts.AssertNonEmpty(src);
+            Contracts.Assert(0 < count && count <= src.Length);
+            Contracts.AssertNonEmpty(indices);
+            Contracts.Assert(count <= indices.Length);
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(count < dst.Length);
+
+            AddScale(a, new Span<float>(src), new Span<int>(indices, 0, count), new Span<float>(dst));
+        }
+
+        public static void AddScale(float a, float[] src, int[] indices, float[] dst, int dstOffset, int count)
+        {
+            Contracts.AssertNonEmpty(src);
+            Contracts.Assert(0 < count && count <= src.Length);
+            Contracts.AssertNonEmpty(indices);
+            Contracts.Assert(count <= indices.Length);
+            Contracts.AssertNonEmpty(dst);
+            Contracts.Assert(0 <= dstOffset && dstOffset < dst.Length);
+            Contracts.Assert(count < dst.Length - dstOffset);
+
+            AddScale(a, new Span<float>(src), new Span<int>(indices, 0, count), 
+                    new Span<float>(dst, dstOffset, dst.Length - dstOffset));
+        }
+
+        private static void AddScale(float a, Span<float> src, Span<int> indices, Span<float> dst)
+        {
+            if (Sse.IsSupported)
+            {
+                SseIntrinsics.AddScaleSU(a, src, indices, dst);
+            }
+            else
+            {
+                for (int i = 0; i < indices.Length; i++)
+                {
+                    int index = indices[i];
+                    dst[index] += a * src[i];
```

Comment:
Is `src` always considered "dense" here?  and only `dst` is sparse?

Reply:
Yes, you're right.  It seems to me that `src` is always dense, while we only deal with the indexed elements of `dst`. 
 I looked up the native code for this software fallback: https://github.com/dotnet/machinelearning/blob/master/src/Native/CpuMathNative/Sse.cpp#L1684-L1700.